Add libraries support

56738 opened the pull request:
Adds `library`, `bukkitLibrary` and `bungeeLibrary` dependency configurations.

Specifying a dependency like this:
```kotlin
dependencies {
    library(kotlin("stdlib"))
}
```
generates this output:
```yaml
libraries:
 - org.jetbrains.kotlin:kotlin-stdlib:1.5.20
```
Alternatively, `libraries` can be configured manually in the build script:
```yaml
bukkit {
    // ...
    libraries = listOf(
        "org.jetbrains.kotlin:kotlin-stdlib:1.5.20"
    )
}
```

Closes #11

56738 committed:
Add libraries support

Closes #11

stephan-gh commented:
This seems really nice, thanks for this! Could you please rebase this and fix the conflict after I merged #10?

56738 commented:
Rebased and replaced `getByName` with `maybeCreate` because the `compileClasspath` configuration doesn't exist before this plugin is applied in some situations.

stephan-gh commented:
Thanks!

Hmm, I'm not sure the `maybeCreate("compileClasspath")` is a very good idea. I think what we should rather do is to only update the compileClasspath configuration from this part of the code:

https://github.com/Minecrell/plugin-yml/blob/5e1f0bc5774e5cd0cf8edccb363a365752cc385e/src/main/kotlin/net/minecrell/pluginyml/PlatformPlugin.kt#L63-L67

Which is delayed until the JavaPlugin has been applied. Perhaps something like 

```kotlin
configurations.getByName(compileClasspathConfigurationName).extendsFrom(platformLibrary);
```

`compileClasspathConfigurationName` refers to `SourceSet.getCompileClasspathConfigurationName()` here.

Also, there seems to be a bit of duplication for the `createConfiguration` code. I think you could just put that into the `PlatformPlugin` like `project.configurations.maybeCreate("${platformName}Library")`.

To avoid it for Nukkit, perhaps add the `createConfiguration()` implementation as `open fun` to `PlatformPlugin` and then keep the override that returns null for Nukkit. Does that make sense?

56738 committed:
Refactor library configuration

56738 commented:
Thanks, that makes sense and looks much better.

56738 committed:
Add missing null check

stephan-gh commented on src/main/kotlin/net/minecrell/pluginyml/PlatformPlugin.kt:
```diff
@@ -37,6 +38,11 @@ abstract class PlatformPlugin<T : PluginDescription>(private val platformName: S
 
+    protected open fun createConfiguration(project: Project): Configuration? {
+        val library = project.configurations.maybeCreate("library")
+        return project.configurations.create("${platformName}Library").extendsFrom(library)
```
You probably want 

```suggestion
        return project.configurations.create("${platformName.decapitalize()}Library").extendsFrom(library)
```

here or it will create the configurations like `BukkitLibrary` instead of `bukkitLibrary`.

56738 committed:
Decapitalize platform library configuration

stephan-gh commented:
One last thing, could you check if the `libraries` property in Spigot/BungeeCord expects us to provide all transitive dependencies?

For example, let's say we do:

```
library("com.fasterxml.jackson.core:jackson-databind:2.12.4")
```

With the current code you'd generate:

```yaml
libraries:
 - com.fasterxml.jackson.core:jackson-annotations:2.12.4
 - com.fasterxml.jackson.core:jackson-core:2.12.4
 - com.fasterxml.jackson.core:jackson-databind:2.12.4
```

I wonder if Spigot/BungeeCord will download the transitive dependencies of jackson-databind (jackson-core and jackson-annotations) automatically. Would it be enough to say this?

```yaml
libraries:
 - com.fasterxml.jackson.core:jackson-databind:2.12.4
```

56738 committed:
Only include top-level dependencies in libraries

56738 commented:
You're right, the library loader resolves and downloads transitive dependencies.

stephan-gh commented:
Thanks for your contribution, this is a nice way to integrate this functionality IMO. :)
